### PR TITLE
Add premium content button block for login and subscribe buttons

### LINKED ISSUE #41850: Premium Content: Decouple buttons from the block
Currently, the buttons inside the premium content block are "fake" buttons that are only displayed and not really editable.

The ideal solution would:
- Decouple "Log In" button as a standalone component that works outside the block
- Decouple the "Subscribe" button the same way (bonus points if we can reuse recurring payments button to that end)

That would even open up the premium content block to be usable without buttons

@alaczek has pointed out that she needs the following for the page templates:

--------

I was exploring ideas for premium content block patterns this week. I started out in the editor, but immediately hit a roadblock (more on that below), so I turned to Sketch instead and put together some ideas.

The biggest issue, that blocks us from doing anything beyond the simplest layouts (like the default pure text one), is not being able to nest the buttons in other blocks. I talk about this, and a few more suggestions below, starting with the most impactful.

**Blockers:**
* ability to nest the buttons - this would open up options to put them inside a group, layout grid, cover or media & text blocks, which are some of the basic design tools in the editor that we routinely take advantage of.
* align the buttons (left, right, center), to bring parity with the Buttons block.

Here’s a video to illustrate these issues: https://cloudup.com/cSbS-80-MMX

This would enable us to start creating designs like these:
premium content patterns 4
_(some examples are more aspirational because of Guetnberg limitations)_

**We need those two issues addressed before I can make any shippable block patterns for premium content.** 

**For later:**
* be able to style each button separately. Having two identical buttons (two primary actions) next to each other is distracting. We could take a cue from the Buttons block here, where each button has it’s own settings. Ideally we would have all the button styling options available, so not just background and text color, but style variations and border-radius as well.

### PR DESCRIPTION
**Pull request by eoigal**
#### TL;DR

Changes here will offer more flexibility for the Subscribe/Login buttons allowing users:
* Insert a Premium Content block without them.
* Style and format them independently.
* Insert a Premium Content login button anywhere (even outside the Premium Content block).

#### Full summary 

* Defines a [Block Context](https://github.com/WordPress/gutenberg/blob/master/docs/designers-developers/developers/block-api/block-context.md) in the container block which is consumed by the nested blocks (subscriber and non-subscriber view blocks) to know what the container's selected plan ID is and determine if those blocks should be rendered or not.
* Decouple the login and subscribe buttons used in the non-subscriber view of the premium content block.

For the latter, this PR re-uses the existing Recurring Payment block for the subscribe button and registers a new block for the login button. That way, each button can be styled and formatted independently.

Screen Shot 2020-06-11 at 13 35 50

The Recurring Payment block automatically selects the plan selected on the Premium Content block.

The new login button block is implemented as a copy of the [Core Button block](https://github.com/WordPress/gutenberg/tree/master/packages/block-library/src/button) but leaving out the features that are not applicable for our case (such as setting the URL).

AFAIK, the URL needs to be determined URL, so there is a small dynamic render callback that adds the `href` attribute to the block.

### Testing instructions

* Switch to a test site with a Recurring Payments plan created.
* Before applying this branch, make sure you have a post published that contains a Premium Content block using the current FSE version. Try to customize the block as much as possible (i.e. button labels, background and text colors, ...).
* Apply D44043-code or `cd apps/full-site-editing && yarn dev --sync`.
* Visit the published post as a logged out user.
* Make sure the non-subscriber view is displayed correctly (no differences with production).
* Test that the login button, well, logs you in 😄 . 
* Test that the subscribe button opens the recurring payments modal to allow you to subscribe to the blocks plan; and confirm that when you close that modal (after you have subscribed) that you can view the premium content automatically.
* Open the post in the block editor.
* Test that the buttons are rendered ok in the editor (they should've been migrated to the new markup). The only visual difference should be the width (it grows now with content, previously fixed to 200px) and the margin ([we use now the same margin used by the Core Buttons block](https://github.com/Automattic/wp-calypso/pull/42715#issuecomment-642591381)).
* Create a new post and insert a Premium Content block.
* Test that you can remove all buttons and add them again.
* Make sure the subscriber button block (which is a Recurring Payments block) updates its plan automatically whenever you change the selected plan on the Premium Block.
* Publish the post.
* Verify the subscribe/log in button continue behaving correctly for non-subscribers.
* Create a new post and insert a Premium Content buttons block.
* Verify you can insert a Recurring Payments, a Premium Content login button or a Core button inside the Premium Content buttons block.
* Verify you can insert a Premium Content login button anywhere (even outside the Premium Content block or the Premium Content Buttons block).
* Publish the post.
* Make sure the Premium Content login button is visible only while logged out.
* Switch to a site under a free plan.
* Make sure neither the Premium Content buttons nor Premium Content login button can be inserted.

Fixes #41850

### CODE REVIEW

**Comment by kwight** [comment]
Sweet 🎉 

> Test that you can remove all buttons and add them again inside the logged out (non-subscriber) view - you shouldn't be able to add a Premium content button to the subscriber view OR outside the Premium content container.

I only saw the inserter within the container. However, it would only insert a Login button – I couldn't figure out how to add back a Subscribe button. I could make multiple Login buttons though.

**Comment by eoigal** [comment]
> While testing on an existing page with the premium content block I did notice that the buttons were no longer available, so we might want to check if we need any block transformations for older versions.

I hadn't considered that. I'll look into adding support for older versions.

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+ * @property { () => void } setTextButtonColor
+ * @property { (attributes: Partial<Attributes>) => void } setAttributes
+ *
+ * @param { Props } props
+ */
+function Edit( props ) {
+    useEffect( () => {
+        setButtonClasses();
+        setCustomButtonColors();
+        setBlockContext();
```
[the remark this answers is not in the document]
Comment (eoigal) [reply apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js:170]: I haven't seen any weirdness but happy to into any suggestions in how to make this less expensive. How would you batch these updates?

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+} from '@wordpress/block-editor';
+import { get } from 'lodash';
+import classnames from 'classnames';
+
+const { getComputedStyle } = window;
+
+const applyFallbackStyles = withFallbackStyles( ( node, ownProps ) => {
+    const { textButtonColor, backgroundButtonColor } = ownProps;
+    const backgroundColorValue = backgroundButtonColor && backgroundButtonColor.color;
+    const textColorValue = textButtonColor && textButtonColor.color;
```
[the remark this answers is not in the document]
Comment (eoigal) [reply apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js:22]: Neat. I'll try this out and see if it works.

**Comment by mmtr** [comment]
> I only saw the inserter within the container. However, it would only insert a Login button – I couldn't figure out how to add back a Subscribe button. I could make multiple Login buttons though.

You need to open the sidebar so you can choose the button type:

Screen Shot 2020-05-28 at 15 25 04

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (gwwar) [reply apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js:170]: It'd mainly be setting the block attributes with one call vs 3. I suspect we trigger additional work here (eg think of listeners that wait for block changes), but if you aren't seeing any update oddness this is an optional note

**Comment by eoigal** [comment]
I've added a new function in the logged out view block that checks to see if there are any buttons in that blocks innerBlocks content. If none found, we'll dynamically insert them using the block attributes as default values - this means for posts on a older version of the plugin, they should now be able to edit the login and subscribe buttons inside the innerBlocks area and those buttons should have the same attributes as before.

I have to do something similar on the server side, so that when a post on a older version of the plugin is rendering blocks for a published post, the new button blocks will be used.

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/index.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    /* translators: block name */
+    title: __( 'Premium Content Button', 'premium-content' ),
+    /* translators: block description */
+    description: __( 'Premium Content Button.', 'premium-content' ),
+    parent: [ 'premium-content/logged-out-view' ],
+    supports: {
+        html: false,
+    },
+    edit,
+    save: () => null,
```
Comment (mmtr) [inline apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/index.js:108]: Is there any reason why we need to render these buttons server-side? Looking at `premium_content_render_button_block` I think the only button attribute that's retrieved from the server is the URL (via `premium_content_subscription_service()->access_url()`). If we could have that URL in the client, we wouldn't need the dynamic rendering and the implementation could be simplified down to a `InnerBlocks` component containing `core/button` blocks.

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    const buttonStyle = { border: 'none', backgroundColor, color };
+    const buttonClasses = getButtonClasses();
+
+    return [
+        isSelected && blockControls( props ),
+
+        isSelected && inspectorControls( props ),
+
+        <div className={ props.className }>
+            <div className="wp-block-button premium-content-logged-out-view-button">
```
Comment (mmtr) [inline apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js:239]: If we could save this block in the client, it should be possible to re-use the `core/button` blocks instead of reimplementing them here:

```
const TEMPLATE = [
  [ 'core/button', { 
    text: attributes.buttonType === 'login' ? 'Log in' : 'Subscribe',
    url: attributes.buttonType === 'login' ? accessUrl : subscribeUrl, // We need to expose those URLs in the client somehow.
  } ]
];
```

```
<InnerBlocks
  template={ TEMPLATE }
  templateLock="insert"
/>
```

With `templateLock="insert"` we make sure blocks cannot be inserted or removed, but existing blocks provided by the template can be updated.

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/index.js (hunk 3 above)
Comment (eoigal) [reply apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/index.js:108]: That's a good point and I'm not sure to be honest. We also do authentication to determine what blocks are rendered server side but perhaps this isn't necessary for these buttons. I'll look into it.

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js [hunk 5]
```diff
@@ -0,0 +1,257 @@
+import { __ } from '@wordpress/i18n';
+import { useEffect } from '@wordpress/element';
+import { compose } from '@wordpress/compose';
+import { Button, ButtonGroup, PanelBody, withFallbackStyles } from '@wordpress/components';
+import {
+    BlockAlignmentToolbar,
```
[the remark this answers is not in the document]
Comment (eoigal) [reply apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js:6]: Ugh, sorry about that. I'm not sure why phpstorm does that.

**Comment by eoigal** [comment]
I think this is ready to be reviewed again. From my testing this PR works with posts with older versions of premium content blocks, like [here](https://eointestfse.wordpress.com/2020/05/11/premium-content/).

**Comment by gwwar** [comment]
Thanks @eoigal we're getting closer! Manually testing this, existing premium content blocks appear to be preserved, but I'm seeing doubles on new inserts:

Screen Shot 2020-05-29 at 2 44 36 PM

@eoigal is out till Wednesday, so this PR can be taken over by @Automattic/earn or @Automattic/serenity if folks would like to help move this along.

**Comment by mmtr** [comment]
I'll gave this a spin today 👍

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/index.js (hunk 3 above)
Comment (mmtr) [reply apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/index.js]: Right, doesn't seem to be that trivial. I also noted that using `core/button` blocks will expose more settings than we need (such as replacing the link or defining the `_target` attribute), so it's definitely better if we implement our own blocks. 

Even though, it'd be great to move part of the rendering to the client, but that can be done later.

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js (hunk 4 above)
Comment (mmtr) [reply apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js]: Realized it's better to implement our own blocks rather than re-using `core/block`: https://github.com/Automattic/wp-calypso/pull/42715#discussion_r433197154

**Comment by mmtr** [comment]
> I'm seeing doubles on new inserts

Took a look at that today but couldn't get a proper way of fixing it. The issue is that [`insertButtonBlocksIfMissing`](https://github.com/Automattic/wp-calypso/blob/6a468372c46e8868a3befa2ca3b9a154437ad161/apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/logged-out-view/edit.js#L119) is fired too soon and the [default inner blocks](https://github.com/Automattic/wp-calypso/blob/6a468372c46e8868a3befa2ca3b9a154437ad161/apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/logged-out-view/edit.js#L85-L92) have not being inserted yet, causing the logic to think the block doesn't contain them. Folks are welcome to continue investigating it.

**Comment by n3f** [comment]
> I've added a new function in the logged out view block that checks to see if there are any buttons in that blocks innerBlocks content. If none found, we'll dynamically insert them using the block attributes as default values

Does this mean a user can't create a block without buttons?

**Comment by gwwar** [comment]
> Does this mean a user can't create a block without buttons?

With the previous implementation, I don't think it was possible to not render the buttons. 

If we decouple them from the server callback with this PR, it should be possible.

I can help take a look on Tues or Wed if folks are still stuck.

**Comment by mmtr** [comment]
Tried a different approach in b7a3a72 so the buttons are (almost) completely decoupled from the server callback, effectively addressing the duplicated buttons issues reported by @gwwar. 

Summary of the changes introduced in b7a3a72:
- Buttons are now saved to the post content duplicating most of the existing `core/button` block logic (so they behave and look the same).
- For the same reasons, buttons are now wrapped in a `core/buttons` block.
- The only thing that is not saved to the post content are the URLs. They are not exposed to the client, so they are added by the server-side rendering callback.
- There is a new deprecation for the `premium-content/logged-out-view` block that will migrate old blocks to the new markup.
- Existing published posts will continue rendering the buttons server-side.

There is still a few of issues we have to fix before merging this so the behavior continues matching  production (namely opening the subscribe dialog with `thickbox` or preserving colors when migrating old blocks), but wanted to push the current state of the new approach in case folks have feedback.

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/color-edit.js [hunk 6]
```diff
@@ -0,0 +1,211 @@
+// The code in this file is copied entirely from https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/button/color-edit.js
```
Comment (mmtr) [inline apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/color-edit.js:1]: Code copied from Gutenberg, feel free to skip it during review.

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/color-props.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
+// The code in this file is copied entirely from https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/button/color-props.js
```
Comment (mmtr) [inline apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/color-props.js:1]: Code copied from Gutenberg, feel free to skip it during review.

**Comment by mmtr** [comment]
Alright, this should be ready for review. Folks are welcome to push changes (or even revert the new approach) if needed.

**Comment by eoigal** [comment]
This looks amazing, thanks for taking this up while I was away Miguel. I prefer your approach as the customization options are more inline with what are already available on Gutenberg and it generally looks more robust.

One thing I noticed however is that we aren't passing the button attributes to the block editor when rendering buttons on a post created with an older version of premium content block.

**Old published post:**
Screenshot 2020-06-04 at 14 39 35

**Post in block editor:**
Screenshot 2020-06-04 at 14 39 27

**Comment by gwwar** [comment]
If things look good don't forget to approve @eoigal ! I'll give this branch a spin too.

**Comment by kwight** [comment]
I ran through the instructions and everything worked great 🎉 I hit two weird spots, but I'll verify and see if they're also prod issues first.

I noticed when using the mini-inserter, we get the choice of a Button or a Premium Content Button. The Button has an additional option of Fill and Outline styles, while the Premium Content ones don't. I'm also curious why we don't default a new Premium Content to either Log in or Subscribe; they're neither on insertion:

Screen Shot 2020-06-04 at 10 25 37 AM

And I think it should be "Log in" (with a space), as an action, like "Subscribe"?

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/color-edit.js (hunk 6 above)
Comment (gwwar) [reply apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/color-edit.js:1]: Maybe leave an issue for follow up? This feels like something we could try exporting as a npm package from the core pacakge if there's general reuse value here.

**Comment by mmtr** [comment]
> I noticed when using the mini-inserter, we get the choice of a Button or a Premium Content Button. 

Yep, this is both unavoidable and intended. Unavoidable because we use a `core/buttons` block for wrapping our subscribe/login buttons, and the `core/button` block [can be inserted on it](https://github.com/WordPress/gutenberg/blob/6a90f56975ac6d1a7f3e45df35ee21bab04df8fb/packages/block-library/src/button/block.json#L4-L6). But also intended because users may want to insert other buttons that are neither a subscribe nor a login button.

> The Button has an additional option of Fill and Outline styles, while the Premium Content ones don't. 

Investigated this and confirmed we can re-use the `core/button` styles for our premium content button blocks. Added in 0e9b5ad.

> I'm also curious why we don't default a new Premium Content Button to either Log in or Subscribe; they're neither on insertion

Definitely a bug! Fixed in 30ef4d2.
 
> And I think it should be "Log in" (with a space), as an action, like "Subscribe"?

Good catch! Also fixed in 30ef4d2.

**Comment by mmtr** [comment]
> Block attributes serialize to different markup than what was saved previously.

Nicely spotted @gwwar. Should've been fixed in 93f7ff3

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/color-edit.js (hunk 6 above)
Comment (mmtr) [reply apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/color-edit.js:1]: Just found out there is a plan for removing these files in favor of using the new `color` and `style` support flags. It was actually done in https://github.com/WordPress/gutenberg/pull/21266 but reverted in https://github.com/WordPress/gutenberg/pull/21923 due to regressions. 

Will investigate if we can use those support flags to reduce code here.

Comment (mmtr) [reply apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/color-edit.js]: Was able to get rid of those duplicate files in favor of using [the `color` hooks](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/hooks/color.js) in 750d3d8 🎉

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+				style={ {
+					borderRadius: borderRadius ? borderRadius + 'px' : undefined,
+				} }
+			/>
+			<InspectorControls>
+				<PanelBody title={ __( 'Button Type', 'premium-content' ) }>
+					<ButtonGroup aria-label={ __( 'Button Type', 'premium-content' ) }>
+						{ Object.entries( BUTTON_TYPES ).map( ( [ buttonType, buttonText ] ) => (
+							<Button
+								isLarge
```
Comment (kwight) [inline apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/button/edit.js:68]: I think we need a key here; I'm seeing `Warning: Each child in a list should have a unique "key" prop.` in the console.

```suggestion
								isLarge
								key={ buttonText }
```

**Comment by kwight** [comment]
Sweet – from what I can tell, spelling, styles, default button type and serialization issues have all been addressed. I added one comment to deal with a console warning, but otherwise, giving the thumbs up 👍

**Comment by gwwar** [comment]
Looking really close. I verified that serialization cases look good. 

Some things I noticed:
* Button Styling will be different on edit (after it's converted), the deprecated transform doesn't appear to trigger changes unsaved changes (though this might be intentional by Gutenberg).
* I also noticed was that we could insert a premium content button, but this looks like it's **only** an option with the immediate parent being a premium content block. 

> ability to nest the buttons - this would open up options to put them inside a group, layout grid, cover or media & text blocks, which are some of the basic design tools in the editor that we routinely take advantage of.

@artpi we wanted to be able to have this button be a child of group and other containers correct?

So for example we'd show this in the inserted provided that a parent was a premium content block. Eg `Premium Content > Group > Premium Content Block` `Premium Content > Cover > Premium Content Block` vs just `Premium Content > Premium Content Block`

**Comment by mmtr** [comment]
> I also noticed was that we could insert a premium content button, but this looks like it's **only** an option with the immediate parent being a premium content block.

It should be possible to insert it in the non-subscriber view only. And to clarify, this is intentional as described in https://github.com/Automattic/wp-calypso/issues/41850:

> ability to nest the buttons - this would open up options to put them inside a group, layout grid, cover or media & text blocks, which are some of the basic design tools in the editor that we routinely take advantage of.

**Comment by mmtr** [comment]
> Button Styling will be different on edit (after it's converted), the deprecated transform doesn't appear to trigger changes unsaved changes (though this might be intentional by Gutenberg).

I don't totally follow you on this, could you clarify a bit further? 

Buttons keep having the same styling for me when migrating an old block to the new markup. I tried different scenarios and they all worked as expected:
- Reload editor before updating (new markup has not been saved before reload, so the deprecated markup will be migrated to the new one upon reloading).
- Reload editor after updating (new markup is saved, preserving old styles).

**Comment by gwwar** [comment]
> It should be possible to insert it in the non-subscriber view only.

To clarify, when I last tested:
1. Try inserting a premium content block
2. Within the logged out view insert a group or other container like cover
3. Try searching for the premium content button in the inserter
Expected:  it appears as an option
Actual: It doesn't appear. Only as a direct child to the logged out view.

**Comment by gwwar** [comment]
> Buttons keep having the same styling for me when migrating an old block to the new markup

Note that I've sandboxed in both images:

| Existing Button Without Patch Applied | Existing Button With Patch Applied |
| ------------- | ------------- |
| Screen Shot 2020-06-08 at 3 17 16 PM | after |

**Comment by mmtr** [comment]
> To clarify, when I last tested:
> 
> 1. Try inserting a premium content block
> 2. Within the logged out view insert a group or other container like cover
> 3. Try searching for the premium content button in the inserter
>    Expected:  it appears as an option
>    Actual: It doesn't appear. Only as a direct child to the logged out view.

Gotcha, thanks for clarifying. Seems [the `parent` option](https://github.com/WordPress/gutenberg/blob/master/docs/designers-developers/developers/block-api/block-registration.md#parent-optional) when registering a block requires the block to be a direct child.

In order to be able to have the button be an any-level descendant of any other block, we'd need to dynamically change one of those:
* The `InnerBlocks`'s allowed blocks of a descendant block. i.e. `premium-content/button` would be allowed in [`core/buttons`](https://github.com/WordPress/gutenberg/blob/e0b1d1d0121628e24033d3245bbffd3ddc64babd/packages/block-library/src/buttons/edit.js#L28) if the Core buttons block has a `premium-content/logged-out-view` ancestor, and disallowed otherwise.
* The `parent` property of the button block. i.e. `premium-content/button` could have a `core/buttons` block only if the Core buttons block has a `premium-content/logged-out-view` ancestor.

Unfortunately, there is no way in Gutenberg to achieve that currently (see https://github.com/WordPress/gutenberg/issues/14515 or https://github.com/WordPress/gutenberg/issues/15682). I'd suggest to go ahead with this limitation (buttons can be only direct children of the non-subscriber view) and try to follow up with an upstream patch in Gutenberg so allowed nested blocks can be filtered dynamically.

On a related note, I slightly changed the behavior in 05b1bb5. We were previously defining `core/buttons` as parent of `premium-content/button`, but thanks to allowed blocks limitations learn up, I realized that that was making possible to insert a premium content subscribe button outside the premium content block (by inserting a Core buttons block). To have things more constrained, in 05b1bb5 I registered a new `premium-content/buttons` block that is only insertable in a `premium-content/logged-out-view` block and allows to nest both `premium-content/button` and `core/button` blocks.

**Comment by mmtr** [comment]
> > Buttons keep having the same styling for me when migrating an old block to the new markup
> 
> Note that I've sandboxed in both images, but let me know if other folks can't repro:
> 
> Existing Button Without Patch Applied	Existing Button With Patch Applied
> Screen Shot 2020-06-08 at 3 17 16 PM	after

Oh right, likely your theme is defining some styles for the `.wp-block-button` class which is not added when using the new colors hook (and the reason why WordPress/gutenberg#21266 was reverted).

Let me see if I can find a work-around. Otherwise, we might need to bring back the duplicated `ColorEdit` component and wait until Gutenberg figures out a way of using buttons with the colors hooks without regressions.

**Comment by mmtr** [comment]
> Otherwise, we might need to bring back the duplicated `ColorEdit` component and wait until Gutenberg figures out a way of using buttons with the colors hooks without regressions.

Finally had to take this approach. Themes rely a lot on the `wp-block-button` wrapper class which is incompatible with the `__experimentalColor` hook (this is already a known issue in Gutenberg).

That has made this PR a bit bigger, but part of the code is duplicated. Let me know if folks prefer to split work here into more reasonable chunks.

**Comment by mmtr** [comment]
> New Buttons are only addable from the buttons container

Actually, this might be a blocker since the Recurring Payments is not a new block and should be possible to insert it outside the Buttons group. Will look into this before merging.

> Minor Styling Differences for an Existing Premium Content Block

> It looks like the new buttons have some top margin that the originals do not

Will double check this but it could be intentional. I removed all the previous margin styles so buttons get the default styling defined by themes for the `wp-block-buttons`, `wp-block-button` and `wp-block_button__link` classes.

Do you get the same styling if you insert a regular Buttons block with a few Core Button blocks inside?

**Comment by mmtr** [comment]
> would they need to know to use a Recurring Payments button, and then set the plan appropriately?

Yes and no. They'll need to add back the Recurring Payments button, but the plan is set automatically to the same one that has been selected in the Premium Content block.

**Comment by mmtr** [comment]
> Actually, this might be a blocker since the Recurring Payments is not a new block and should be possible to insert it outside the Buttons group. Will look into this before merging.

Fixed in 3e40b47. I added a new `premium-content/buttons` block as a container for the `premium-content/login-button`, `jetpack/recurrring-payments` and `core/button` blocks, but all of them can still be inserted (and discovered) outside this container.

Screen Shot 2020-06-11 at 13 38 30

Screen Shot 2020-06-11 at 13 40 06

> Do you get the same styling if you insert a regular Buttons block with a few Core Button blocks inside?

Double checked this and verified that all combinations share the same styling.

Screen Shot 2020-06-11 at 13 46 24

**Comment by davemart-in** [comment]
Excellent work @mmtr!

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/login-button/color-edit.js [hunk 9]
```diff
@@ -0,0 +1,213 @@
+// The code in this file is copied entirely from https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/button/color-edit.js
```
Comment (mmtr) [inline apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/login-button/color-edit.js:1]: Code copied from Gutenberg, feel free to skip during review (it'd be possible to get rid of this once the Core Button block can use the new colors hook: https://github.com/WordPress/gutenberg/issues/22700).

File: apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/login-button/color-props.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
+// The code in this file is copied entirely from https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/button/color-props.js
```
Comment (mmtr) [inline apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/login-button/color-props.js:1]: Code copied from Gutenberg, feel free to skip during review (it'd be possible to get rid of this once the Core Button block can use the new colors hook: https://github.com/WordPress/gutenberg/issues/22700).

**Comment by eoigal** [comment]
Great work Miguel! I'm seeing a couple minor issues however with existing buttons styles (posts with older version of premium content block) in the block editor. It appears that the button styles aren't being passed properly from the logged out view attributes.

I think the block editor view should closely match what the block looks like when published. We may need to change the styles in the block editor to fix the spacing and button widths to do that.

Screenshot 2020-06-11 at 14 42 00
Screenshot 2020-06-11 at 14 41 31

**Comment by mmtr** [comment]
Thanks for testing @eoigal!

> Screen Shot 2020-06-11 at 16 23 53

Good catch! I think this is a separate issue in the Recurring Payments block. For some reason is ignoring the `customTextButtonColor` attribute when rendering the submit button but it's picking twice `customBackgroundButtonColor`.

https://github.com/Automattic/jetpack/blob/266889b064113bd4afd3d1f32d503290330787e0/extensions/blocks/recurring-payments/edit.jsx#L520-L521

If you try to change the text color of a Recurring Payments block you'll see that is not updated. Likely it can be solved by replacing one of the duplicated `customBackgroundButtonColor` attrs with `customTextButtonColor`. Will look into it, but it'll be a separate PR since the Recurring Payments block lives in Jetpack.

> Screen Shot 2020-06-11 at 16 28 06

This is intended in order to align the styling of the subscribe/login buttons with the styling used by the Core buttons. The new `premium-content/buttons` and `premium-content/login-button` blocks are re-using the `wp-block-buttons` and `wp-block-button` classes so we can benefit from Core work and decrease our maintenance cost.

If you add a Core buttons block to your post, you should see that their width is adjusted to the label and they are separated from the content around by the same distance.

Screen Shot 2020-06-11 at 16 40 38

However, I happy to disagree and commit if folks think that keeping a separate set of styles is important.

> **Published post**
> Screenshot 2020-06-11 at 14 41 31

That's weird. After re-publishing the post once it's been migrated in the editor, I got a styling that's similar to the editor:

Screen Shot 2020-06-11 at 16 45 37

Did you re-publish your post containing existing blocks?

Note that in my screenshot above, the Premium Content buttons are not separated by the same distance, but that's an theme issue (Maywood) caused by a very restricted CSS ruleset: top/bottom margins are only applied to direct children of the entry content:

Screen Shot 2020-06-11 at 16 47 18

**Comment by mmtr** [comment]
> If you try to change the text color of a Recurring Payments block you'll see that is not updated. Likely it can be solved by replacing one of the duplicated `customBackgroundButtonColor` attrs with `customTextButtonColor`.

Should be fixed in https://github.com/Automattic/jetpack/pull/16139

**Comment by eoigal** [comment]
I can't approve as I'm still an author but I'm happy to see this merged. I just wanted to highlight the spacing issues in case they were missed but I agree that we'll benefit by keeping as close to core buttons as possible. This is a great improvement to the block Miguel, nice work 👍

**Comment by gwwar** [comment]
Nice @mmtr I verified that the buttons can be inserted from the new container and directly within. 

## Some Misc Things I Found

### Additional Whitespace in Editor, different mix-widths on buttons.

| before | after |
| ------------- | ------------- |
| before | after | 

### Published View Has Extra Styles not reflected in Editor?

| before | after |
| ------------- | ------------- |
| before-published | Screen Shot 2020-06-11 at 2 06 36 PM |

Login button in this case I think is getting set with `is-style-outline` in the `Libretto` theme
login

Subscription/recurring payments button doesn't appear to have this issue

### Recurring Payments Button Appears in Subscriber View When Outside Premium Content Block

In order for it to act as a decoupled subscribe button vs contribute and see this button all the time, I think we need a boolean setting or similar to toggle this behavior. Eg imagine if the buttons below said "Subscribe" instead of $X.XX contribution

Screen Shot 2020-06-11 at 2 11 11 PM

### When inserting premium content buttons (container) w/ multiple plans, subscribe text is replaced with X contribution

| insert | plan select |
| ------------- | ------------- |
| Screen Shot 2020-06-11 at 2 12 16 PM | Screen Shot 2020-06-11 at 2 12 21 PM |

**Comment by mmtr** [comment]
Thanks @gwwar for testing again.

> Additional Whitespace in Editor, different min-widths on buttons.

This is expected as this PR seeks to keep the Premium Content buttons as close as possible to Core buttons. See https://github.com/Automattic/wp-calypso/pull/42715#issuecomment-642709325

> Published View Has Extra Styles not reflected in Editor?

Core buttons have this issue too, so likely a theme issue or maybe a Core issue (will investigate and open an upstream issue if needed). However, I don't get the `is-style-outline` class. If you replicate the same behavior on a Core button, do you get the class on it as well?

Jun-12-2020 11-46-00

> Recurring Payments Button Appears in Subscriber View When Outside Premium Content Block

Right, but I don't think that's a new behavior introduced on this PR. You currently can insert a Recurring Payments block in the subscriber view (using the same plan as the wrapping Premium Content block) when testing on production and it will be always visible for subscribers. I think this is fine to address in a follow up.

> When inserting premium content buttons (container) w/ multiple plans, subscribe text is replaced with X contribution

Yeah, unfortunately the Recurring Payments block overwrites the button label when a plan is selected and we cannot disable the behavior within the Premium Content block. We'll need to open a PR in Jetpack for handling this, but I think this is fine to address in a follow up as well.

---

Given that nothing above seems a real blocker to me, I'm going to merge this. This won't immediately affect production, since it will require a new FSE release, so we have time to follow up in case we missed something critical.

**Comment by mmtr** [comment]
> > Published View Has Extra Styles not reflected in Editor?
> 
> Core buttons have this issue too, so likely a theme issue or maybe a Core issue (will investigate and open an upstream issue if needed). 

Turned out to be an issue in the Recurring Payments block. Should be fixed by https://github.com/Automattic/jetpack/pull/16149.
 
> > When inserting premium content buttons (container) w/ multiple plans, subscribe text is replaced with X contribution
> 
> the Recurring Payments block [overwrites the button label](https://github.com/Automattic/jetpack/blob/72a90143f45055aa5be1ea6cedd1d91e5ae6739e/extensions/blocks/recurring-payments/edit.jsx#L343-L363) when a plan is selected and we cannot disable the behavior within the Premium Content block.

Handled in https://github.com/Automattic/jetpack/pull/16148.

**Comment by gwwar** [comment]
Kudos for pushing this one through @mmtr !
